fgpyo/io/io.py

```diff
@@ -0,0 +1,175 @@
+"""
+IO
+-------
+Module for reading and writing files.
+
+The functions in this module make is easy to:
+* check if a file exists and is writeable
+* check if a file and its parent directories exist and are writeable
+* check if a file exists and is readable
+* check if a path exists and is a directory
+* open an appropriate reader or writer based on the file extension
+* writitng items to a file
+* reading lines from a file
+~~~~~~~~
+Examples:
+.. code-block:: python
+   >>> import fgpyo.io.io as IO
+   >>> from pathlib import Path
+   Assert that a path exist and are readable
+   >>> path_flath: Path = Path("example.txt")
+   >>> path_compressed: Path = Path("example.txt.gz")
+   >>> IO.path_is_readable(path_flat)
+   AssertionError: Cannot read non-existent path: example.txt
+   >>> IO.path_is_readable(compressed_file)
+   AssertionError: Cannot read non-existent path: example.txt.gz
+   Write to and read from path
+   >>> write_lines(path = path_flat, lines_to_write=["flat file", 10])
+   >>> write_lines(path = path_compressed, lines_to_write=["gzip file", 10])
+   Read lines from a path into a generator
+   >>> read_lines(path = path_flat)
+   ['flat file', '10']
+   >>> read_lines(path = path_compressed)
+   ['gzip file', '10']
+"""
+
+import gzip
+import io
+import os
+from pathlib import Path
+from typing import IO
+from typing import Any
+from typing import Generator
+from typing import Iterable
+from typing import Iterator
+from typing import Set
+from typing import TextIO
+from typing import Union
+from typing import cast
+
+COMPRESSED_FILE_EXTENSIONS: Set[str] = {".gz", ".bgz"}
+
+
+def path_is_readable(path: Path) -> None:
```

Comment:
For any functions that assert like this please name them `assert_<expected-state>`.